Src/Base/AMReX_CuptiTrace.cpp

```diff
@@ -0,0 +1,294 @@
+#include <AMReX_CuptiTrace.H>
+#ifdef AMREX_USE_CUPTI
+#include <AMReX_Print.H>
+#include <stdio.h>
+#include <map>
+#include <cuda.h>
+#include <cupti.h>
+
+
+namespace amrex {
+
+std::vector<CUpti_Activity_Userdata*> activityRecordUserdata;
+
+
+void CUPTIAPI
+bfrRequestCallback (uint8_t* *bfr, size_t* size, size_t* maxNumRecords) noexcept
+{
+  // Allocate a buffer for use by CUPTI; activity records are stored in the buffer
+  uint8_t* buffer = (uint8_t*) malloc(BFR_SIZE + ALIGNMENT);
+  if (buffer == NULL) {
+    amrex::Print() << "Error: CUPTI requested a buffer but memory cannot be allocated\n";
+    exit(EXIT_FAILURE);
+  }
+
+  // Return buffer and buffer size
+  *bfr = ALIGN_BFR(buffer, ALIGNMENT);
+  *size = BFR_SIZE;
+
+  // Controls maximum number of records to be placed in buffer; setting to zero
+  // fills the buffer with as many records as possible, setting to n > 0 fills
+  // with n records before buffer is returned
+  *maxNumRecords = 0;
+}
+
+void CUPTIAPI
+bfrCompleteCallback (CUcontext ctx, uint32_t streamId, uint8_t* bfr,
+		     size_t size, size_t validSize) noexcept
+{ 
+  CUptiResult status;
+  CUpti_Activity* record = NULL;
+
+  if (validSize > 0) {
+    do {
+      status = cuptiActivityGetNextRecord(bfr, validSize, &record);
+      if (status == CUPTI_SUCCESS) {
+	CUpti_Activity_Userdata* recordUserData = new CUpti_Activity_Userdata();
+	CUpti_ActivityKernel4* kernel = (CUpti_ActivityKernel4*) record;
+
+	// Save record data
+	recordUserData->setRecord(record);
```

Comment:
I think `record` is a pointer that points to somewhere in the user provided buffer `bfr`.   Because `bfr` gets freed later in this function, `record` stored in `recordUserData` will become a dangling pointer.

Reply:
I've removed `record` as an object saved to `recordUserData`, now it just gathers the needed information during the buffer completion callback, and the dangling pointer is cleaned up after `bfr` is freed.